File: chap1/sec1.3/TransportFEUpwind.m
function [u,x,t]=TransportFEUpwind(f,u0,ua,a,b,J,T,N);
% TRANSPORTFEUPWIND solves the 1d transport equation 
%   [u,x,t]=TransportFEUpwind(f,u0,ua,a,b,J,T,N); solves the 1d
%   transport equation u_t+u_x=f in (a,b)x(0,T) with right hand side
%   function f and given intial condition u0, given boundary
%   condition ua at a using upwind finite differences in space with
%   J+1 mesh points and forward Euler in time up to T using N
%   meshpoints in time.

dx=(b-a)/J;
dt=T/N;
x=(a:dx:b)';
t=(0:dt:T);
u(1,1:N+1)=ua(t);
u(1:J+1,1)=u0(x);
for n=1:N,
  u(2:J+1,n+1)=u(2:J+1,n)-dt/dx*(u(2:J+1,n)-u(1:J,n))+dt*f(x(2:J+1),n*dt);
end;

File: chap1/sec1.3/TransportFEUpwind.py
import numpy as np

def transportFEUpwind(f, u0, ua, a, b, J, T, N):
    """
    Solves the 1D transport equation u_t + u_x = f using upwind finite differences
    in space and forward Euler in time.

    Parameters
    ----------
    f : function
        Right-hand side function f(x, t).
    u0 : function
        Initial condition function u0(x).
    ua : function
        Boundary condition function ua(t) at x = a.
    a : float
        Left boundary of the spatial domain.
    b : float
        Right boundary of the spatial domain.
    J : int
        Number of subintervals in the spatial domain.
    T : float
        Final time of the simulation.
    N : int
        Number of time steps.

    Returns
    -------
    u : ndarray
        Solution array of shape (J+1, N+1).
    x : ndarray
        Spatial grid points.
    t : ndarray
        Time grid points.
    """
    dx = (b - a) / J
    dt = T / N
    x = np.linspace(a, b, J + 1)
    t = np.linspace(0, T, N + 1)
    u = np.zeros((J + 1, N + 1))
    
    # Apply boundary and initial conditions
    u[0, :] = ua(t)  # Boundary condition at x = a
    u[:, 0] = u0(x)  # Initial condition
    
    # Time-stepping loop
    for n in range(N):
        u[1:J+1, n+1] = (
            u[1:J+1, n]
            - dt / dx * (u[1:J+1, n] - u[0:J, n])
            + dt * f(x[1:J+1], t[n])
        )
    
    return u, x, t
